staging/src/k8s.io/apiserver/pkg/server/genericapiserver.go

```diff
@@ -432,11 +432,20 @@ func (s preparedGenericAPIServer) Run(stopCh <-chan struct{}) error {
 		// - once drained, http Server Shutdown is invoked with a timeout of 2s,
 		//   net/http waits for 1s for the peer to respond to a GO_AWAY frame, so
 		//   we should wait for a minimum of 2s
-		stopHttpServerCh = drainedCh.Signaled()
+		delayedStopOrDrainedCh = drainedCh.Signaled()
 		shutdownTimeout = 2 * time.Second
 		klog.V(1).InfoS("[graceful-termination] using HTTP Server shutdown timeout", "ShutdownTimeout", shutdownTimeout)
 	}
 
+	// pre-shutdown hooks need to finish before we stop the http server
+	preShutdownHooksHasStoppedCh, stopHttpServerCh := make(chan struct{}), make(chan struct{})
+	go func() {
+		defer close(stopHttpServerCh)
+
+		<-delayedStopOrDrainedCh
+		<-preShutdownHooksHasStoppedCh
```

Comment:
I've talked with @p0lyn0mial offline, and he made a good point about the possibility of preshutdown hook blocking forever, we should add a timeout 

```
select {
<-preShutdownHooksHasStoppedCh
<- timeAfter(x)
```

Reply:
[I got it in a different commit](https://github.com/kubernetes/kubernetes/pull/107952/commits/b05606b224a2f2397803d31820ef334cbd86b994)

Reply:
I've incorporated your change in https://github.com/kubernetes/kubernetes/pull/107952/

Reply:
> I've talked with @p0lyn0mial offline, and he made a good point about the possibility of preshutdown hook blocking forever, we should add a timeout

sure, it can be done in a separate followup PR, since pre-shutdown hook blocking forever seems to be orthogonal to what this PR fixes.